TiCDC panic: checkpointTs is greater than resolvedTs

### What did you do?

run cdc changefeed

### What did you expect to see?

no panic

### What did you see instead?


```
[root@bogon bigCluster]# kubectl --kubeconfig kubeconfig.yml -n cdc-kafka-big-cluster-tps-1712340-1-428 logs -p tc-ticdc-1
[WARN] TiCDC server data-dir is not set. Please use `cdc server --data-dir` to start the cdc server if possible.
panic: checkpointTs is greater than resolvedTs

goroutine 743 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x2?, 0x2?, {0x0?, 0x0?, 0xc205230120?})
        go.uber.org/zap@v1.24.0/zapcore/entry.go:198 +0x65
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc17e838410, {0xc23891d800, 0x2, 0x2})
        go.uber.org/zap@v1.24.0/zapcore/entry.go:264 +0x3ec
go.uber.org/zap.(*Logger).Panic(0xc000680c00?, {0x402b215?, 0x0?}, {0xc23891d800, 0x2, 0x2})
        go.uber.org/zap@v1.24.0/logger.go:258 +0x59
github.com/pingcap/log.Panic({0x402b215?, 0xc319be1d70?}, {0xc23891d800?, 0x7bec120?, 0x0?})
        github.com/pingcap/log@v1.1.1-0.20230317032135-a0d097d16e22/global.go:54 +0x8b
github.com/pingcap/tiflow/cdc/owner.(*changefeed).updateStatus(0xc060535340, 0x6215ce22ae8001a, 0x6215ce22ae80019, 0x6215ce22d40000a)
        github.com/pingcap/tiflow/cdc/owner/changefeed.go:1000 +0x192
github.com/pingcap/tiflow/cdc/owner.(*changefeed).tick(0xc060535340, {0x49668e0?, 0xc0fec17668}, 0x0?)
        github.com/pingcap/tiflow/cdc/owner/changefeed.go:500 +0x1f05
github.com/pingcap/tiflow/cdc/owner.(*changefeed).Tick(0xc060535340, {0x49667f0, 0xc0fec17638}, 0xc014826840?)
        github.com/pingcap/tiflow/cdc/owner/changefeed.go:239 +0x2d7
github.com/pingcap/tiflow/cdc/owner.(*ownerImpl).Tick(0xc067d4abd0, {0x7f03fa9b9a30?, 0xc067d49f60}, {0x4936960?, 0xc014826840})
        github.com/pingcap/tiflow/cdc/owner/owner.go:201 +0x46b
github.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run(0xc047d8f760, {0x7f03fa9b9a30, 0xc067d49f60}, 0xc0018f39e0, 0x2faf080, {0x3f7e53a, 0x5})
        github.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:290 +0xfa6
github.com/pingcap/tiflow/cdc/capture.(*captureImpl).runEtcdWorker(0xc000536c40, {0x4966890?, 0xc067d49f60}, {0x7f03eea5af58, 0xc067d4abd0}, {0x4936960, 0xc014826840}, 0x0?, {0x3f7e53a, 0x5})
        github.com/pingcap/tiflow/cdc/capture/capture.go:522 +0x185
github.com/pingcap/tiflow/cdc/capture.(*captureImpl).campaignOwner(0xc000536c40, {0x4966890?, 0xc000d7e6c0})
        github.com/pingcap/tiflow/cdc/capture/capture.go:473 +0x814
github.com/pingcap/tiflow/cdc/capture.(*captureImpl).run.func3()
        github.com/pingcap/tiflow/cdc/capture/capture.go:328 +0x3c
golang.org/x/sync/errgroup.(*Group).Go.func1()
        golang.org/x/sync@v0.2.0/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
        golang.org/x/sync@v0.2.0/errgroup/errgroup.go:72 +0xa5
```

### Versions of the cluster

TiCDC version (execute `cdc version`):

```console
Release Version: v7.1.0-test
Git Commit Hash: 49faafdaa53d54ab817845af7517f90a8fe10623
Git Branch: heads/refs/tags/v7.1.0-test
UTC Build Time: 2023-05-26 06:53:18
Go Version: go version go1.20.3 linux/amd64
Failpoint Build: false
```

--- a/cdc/processor/processor.go
+++ b/cdc/processor/processor.go
@@ -281,15 +281,9 @@ func (p *processor) IsRemoveTableSpanFinished(span tablepb.Span) (model.Ts, bool
 		return 0, false
 	}
 
-	var tableCheckpointTs uint64
-	state, alreadyExist := p.sinkManager.r.GetTableState(span)
-	if alreadyExist {
-		stats := p.sinkManager.r.GetTableStats(span)
-		tableCheckpointTs = stats.CheckpointTs
-	}
-
-	if !alreadyExist {
-		log.Warn("table should be removing but not found",
+	state, ok := p.sinkManager.r.GetTableState(span)
+	if !ok {
+		log.Warn("table has been stopped",
 			zap.String("captureID", p.captureInfo.ID),
 			zap.String("namespace", p.changefeedID.Namespace),
 			zap.String("changefeed", p.changefeedID.ID),
@@ -297,18 +291,18 @@ func (p *processor) IsRemoveTableSpanFinished(span tablepb.Span) (model.Ts, bool
 		return 0, true
 	}
 
+	stats := p.sinkManager.r.GetTableStats(span)
 	if state != tablepb.TableStateStopped {
 		log.Debug("table is still not stopped",
 			zap.String("captureID", p.captureInfo.ID),
 			zap.String("namespace", p.changefeedID.Namespace),
 			zap.String("changefeed", p.changefeedID.ID),
-			zap.Uint64("checkpointTs", tableCheckpointTs),
+			zap.Uint64("checkpointTs", stats.CheckpointTs),
 			zap.Stringer("span", &span),
 			zap.Any("tableStatus", state))
 		return 0, false
 	}
 
-	stats := p.sinkManager.r.GetTableStats(span)
 	if p.redo.r.Enabled() {
 		p.redo.r.RemoveTable(span)
 	}

--- a/cdc/scheduler/internal/v3/agent/table.go
+++ b/cdc/scheduler/internal/v3/agent/table.go
@@ -76,6 +76,7 @@ func (t *tableSpan) getTableSpanStatus(collectStat bool) tablepb.TableStatus {
 func newAddTableResponseMessage(status tablepb.TableStatus) *schedulepb.Message {
 	if status.Checkpoint.ResolvedTs < status.Checkpoint.CheckpointTs {
 		log.Panic("schedulerv3: resolved ts should not less than checkpoint ts",
+			zap.Any("tableStatus", status),
 			zap.Any("checkpoint", status.Checkpoint.CheckpointTs),
 			zap.Any("resolved", status.Checkpoint.ResolvedTs))
 	}
@@ -94,11 +95,13 @@ func newAddTableResponseMessage(status tablepb.TableStatus) *schedulepb.Message 
 
 func newRemoveTableResponseMessage(status tablepb.TableStatus) *schedulepb.Message {
 	if status.Checkpoint.ResolvedTs < status.Checkpoint.CheckpointTs {
+		// TODO: resolvedTs should not be zero, but we have to handle it for now.
 		if status.Checkpoint.ResolvedTs == 0 {
 			// Advance resolved ts to checkpoint ts if table is removed.
 			status.Checkpoint.ResolvedTs = status.Checkpoint.CheckpointTs
 		} else {
 			log.Panic("schedulerv3: resolved ts should not less than checkpoint ts",
+				zap.Any("tableStatus", status),
 				zap.Any("checkpoint", status.Checkpoint.CheckpointTs),
 				zap.Any("resolved", status.Checkpoint.ResolvedTs))
 		}

--- a/cdc/scheduler/internal/v3/agent/table.go
+++ b/cdc/scheduler/internal/v3/agent/table.go
@@ -74,6 +74,11 @@ func (t *tableSpan) getTableSpanStatus(collectStat bool) tablepb.TableStatus {
 }
 
 func newAddTableResponseMessage(status tablepb.TableStatus) *schedulepb.Message {
+	if status.Checkpoint.ResolvedTs < status.Checkpoint.CheckpointTs {
+		log.Panic("schedulerv3: resolved ts should not less than checkpoint ts",
+			zap.Any("checkpoint", status.Checkpoint.CheckpointTs),
+			zap.Any("resolved", status.Checkpoint.ResolvedTs))
+	}
 	return &schedulepb.Message{
 		MsgType: schedulepb.MsgDispatchTableResponse,
 		DispatchTableResponse: &schedulepb.DispatchTableResponse{
@@ -88,6 +93,16 @@ func newAddTableResponseMessage(status tablepb.TableStatus) *schedulepb.Message 
 }
 
 func newRemoveTableResponseMessage(status tablepb.TableStatus) *schedulepb.Message {
+	if status.Checkpoint.ResolvedTs < status.Checkpoint.CheckpointTs {
+		if status.Checkpoint.ResolvedTs == 0 {
+			// Advance resolved ts to checkpoint ts if table is removed.
+			status.Checkpoint.ResolvedTs = status.Checkpoint.CheckpointTs
+		} else {
+			log.Panic("schedulerv3: resolved ts should not less than checkpoint ts",
+				zap.Any("checkpoint", status.Checkpoint.CheckpointTs),
+				zap.Any("resolved", status.Checkpoint.ResolvedTs))
+		}
+	}
 	message := &schedulepb.Message{
 		MsgType: schedulepb.MsgDispatchTableResponse,
 		DispatchTableResponse: &schedulepb.DispatchTableResponse{

--- a/cdc/scheduler/internal/v3/replication/replication_manager_test.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager_test.go
@@ -522,7 +522,10 @@ func TestReplicationManagerBurstBalanceMoveTables(t *testing.T) {
 				AddTable: &schedulepb.AddTableRequest{
 					Span:        span2,
 					IsSecondary: true,
-					Checkpoint:  tablepb.Checkpoint{CheckpointTs: 1},
+					Checkpoint: tablepb.Checkpoint{
+						CheckpointTs: 1,
+						ResolvedTs:   1,
+					},
 				},
 			},
 		},

--- a/cdc/scheduler/internal/v3/replication/replication_manager_test.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager_test.go
@@ -495,7 +495,7 @@ func TestReplicationManagerBurstBalanceMoveTables(t *testing.T) {
 	table2, err := NewReplicationSet(span2, 0, map[string]*tablepb.TableStatus{
 		"1": {
 			Span: span2, State: tablepb.TableStateReplicating,
-			Checkpoint: tablepb.Checkpoint{CheckpointTs: 1},
+			Checkpoint: tablepb.Checkpoint{CheckpointTs: 1, ResolvedTs: 1},
 		},
 	}, model.ChangeFeedID{})
 	require.Nil(t, err)

--- a/cdc/scheduler/internal/v3/replication/replication_set.go
+++ b/cdc/scheduler/internal/v3/replication/replication_set.go
@@ -915,11 +915,27 @@ func (r *ReplicationSet) handleCaptureShutdown(
 func (r *ReplicationSet) updateCheckpointAndStats(
 	checkpoint tablepb.Checkpoint, stats tablepb.Stats,
 ) {
+	if checkpoint.ResolvedTs < checkpoint.CheckpointTs {
+		log.Warn("schedulerv3: resolved ts should not less than checkpoint ts",
+			zap.Any("replicationSet", r),
+			zap.Any("checkpoint", checkpoint))
+
+		// TODO: resolvedTs should not be zero, but we have to handle it for now.
+		if checkpoint.ResolvedTs == 0 {
+			checkpoint.ResolvedTs = checkpoint.CheckpointTs
+		}
+	}
 	if r.Checkpoint.CheckpointTs < checkpoint.CheckpointTs {
 		r.Checkpoint.CheckpointTs = checkpoint.CheckpointTs
 	}
 	if r.Checkpoint.ResolvedTs < checkpoint.ResolvedTs {
 		r.Checkpoint.ResolvedTs = checkpoint.ResolvedTs
+	}
+	if r.Checkpoint.ResolvedTs < r.Checkpoint.CheckpointTs {
+		log.Panic("schedulerv3: resolved ts should not less than checkpoint ts",
+			zap.Any("replicationSet", r),
+			zap.Any("checkpoint", r.Checkpoint.ResolvedTs),
+			zap.Any("resolved", r.Checkpoint.CheckpointTs))
 	}
 	r.Stats = stats
 }

--- a/cdc/scheduler/internal/v3/replication/replication_set_test.go
+++ b/cdc/scheduler/internal/v3/replication/replication_set_test.go
@@ -105,16 +105,25 @@ func TestNewReplicationSet(t *testing.T) {
 				Captures: map[string]Role{
 					"1": RoleSecondary, "2": RolePrimary,
 				},
-				Checkpoint: tablepb.Checkpoint{CheckpointTs: 2},
+				Checkpoint: tablepb.Checkpoint{
+					CheckpointTs: 2,
+					ResolvedTs:   2,
+				},
 			},
 			tableStatus: map[model.CaptureID]*tablepb.TableStatus{
 				"1": {
-					State:      tablepb.TableStatePreparing,
-					Checkpoint: tablepb.Checkpoint{CheckpointTs: 1},
+					State: tablepb.TableStatePreparing,
+					Checkpoint: tablepb.Checkpoint{
+						CheckpointTs: 1,
+						ResolvedTs:   1,
+					},
 				},
 				"2": {
-					State:      tablepb.TableStateReplicating,
-					Checkpoint: tablepb.Checkpoint{CheckpointTs: 2},
+					State: tablepb.TableStateReplicating,
+					Checkpoint: tablepb.Checkpoint{
+						CheckpointTs: 2,
+						ResolvedTs:   2,
+					},
 				},
 			},
 		},

